### PR TITLE
Added axis-aligned IoU loss to VoteNet

### PR DESCRIPTION
**Pull request by THU17cyz**
Added axis-aligned IoU loss code, modified vote_head and updated doc

### CODE REVIEW

File: configs/votenet/README.md [hunk 1]
```diff
@@ -38,3 +38,13 @@ Then you can use the converted checkpoints following [getting_started.md](../../
...
 
 Since test data preparation randomly downsamples the points, and the test script uses fixed random seeds while the random seeds of validation in training are not fixed, the test results may be slightly different from the results reported above.
+
+## IoU loss
+
+Adding IoU loss (simply =1-IoU) boosts VoteNet's performance.
+
+|  Backbone   | Lr schd | Mem (GB) | Inf time (fps) | AP@0.25 |AP@0.5| Download |
+| :---------: | :-----: | :------: | :------------: | :----: |:----: | :------: |
+|    [PointNet++](./votenet_8x8_scannet-3d-18class_withiouloss.py)     |  3x    |4.1||63.81|44.21|/|
```
Comment (ZwwWayne) [inline configs/votenet/README.md:48]: votenet_8x8_scannet-3d-18class_withiouloss.py -> votenet_iouloss_8x8_scannet-3d-18class.py
The config name should follow the convention of config systems, see [doc here](https://mmdetection3d.readthedocs.io/en/latest/config.html).

File: mmdet3d/models/dense_heads/vote_head.py [hunk 2]
```diff
@@ -39,23 +39,26 @@ class VoteHead(nn.Module):
...
+        conv_cfg=dict(type='Conv1d'),
+        norm_cfg=dict(type='BN1d'),
+        objectness_loss=None,
+        center_loss=None,
+        dir_class_loss=None,
+        dir_res_loss=None,
+        size_class_loss=None,
+        size_res_loss=None,
+        semantic_loss=None,
+        iou_loss=None,
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/vote_head.py:60]: comma is unnecessary.

File: mmdet3d/models/dense_heads/vote_head.py [hunk 3]
```diff
@@ -316,7 +324,23 @@ def loss(self,
             dir_class_loss=dir_class_loss,
             dir_res_loss=dir_res_loss,
             size_class_loss=size_class_loss,
-            size_res_loss=size_res_loss)
+            size_res_loss=size_res_loss,
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/vote_head.py:327]: clean comma

File: mmdet3d/models/losses/iou_loss_axis_aligned.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                          reduction='mean'):
+    """Calculate the IoU loss (1-IoU) of two axis aligned bounding boxes.
+
+    Args:
+        center_preds (torch.Tensor): Center predictions with shape [B, N, 3].
+        size_preds (torch.Tensor): Size predictions with shape [B, N, 3].
+        center_targets (torch.Tensor): Center targets (gt) \
+            with shape [B, N, 3].
+        size_targets (torch.Tensor): Size targets with shape [B, N, 3].
+        weight (torch.Tensor or float): Weight of loss.
```
Comment (ZwwWayne) [inline mmdet3d/models/losses/iou_loss_axis_aligned.py:21]: weight (torch.Tensor|float, optional): Weight of loss. Defaults to 1.0.
reduction (str, optional): xxxxx. Defaults to 'mean'.

File: mmdet3d/models/losses/iou_loss_axis_aligned.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        reduction (str): Method to reduce losses.
```
Comment (ZwwWayne) [inline mmdet3d/models/losses/iou_loss_axis_aligned.py:22]: We do not need to define `reduction` head, we can use decorator `@weighted_loss`, see [example here](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/losses/iou_loss.py).

File: mmdet3d/models/losses/iou_loss_axis_aligned.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    Args:
+        reduction (str): Method to reduce losses.
+            The valid reduction method are none, sum or mean.
+        loss_weight (float): Weight of loss.
+    """
+
+    def __init__(
+        self,
+        reduction='mean',
+        loss_weight=1.0,
```
Comment (ZwwWayne) [inline mmdet3d/models/losses/iou_loss_axis_aligned.py:68]: clean comma.

File: mmdet3d/models/losses/iou_loss_axis_aligned.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    ):
+        super(IoULossAxisAligned, self).__init__()
+        assert reduction in ['none', 'sum', 'mean']
+        self.reduction = reduction
+        self.loss_weight = loss_weight
+
+    def forward(self,
+                center_preds,
```
Comment (ZwwWayne) [inline mmdet3d/models/losses/iou_loss_axis_aligned.py:76]: The input of the forward function should be only consistent as those in 2D IoU loss [here](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/losses/iou_loss.py#L353).

File: mmdet3d/models/dense_heads/vote_head.py [hunk 8]
```diff
@@ -318,6 +324,25 @@ def loss(self,
...
+            center_pred = bbox_preds['center']
+            size_pred = torch.clamp(size_pred, 0)
+            pred = torch.cat(
+                [center_pred - size_pred / 2, center_pred + size_pred / 2],
+                dim=-1)
+            target = torch.cat([
+                assigned_center_targets - size_target / 2,
+                assigned_center_targets + size_target / 2
+            ],
+                               dim=-1)
```
Comment (Tai-Wang) [inline mmdet3d/models/dense_heads/vote_head.py:342]: Seems a little bit strange? Is it the result after pre-commit fix?

**Review by Tai-Wang — APPROVED** [review]

File: configs/votenet/README.md [hunk 9]
```diff
@@ -38,3 +38,13 @@ Then you can use the converted checkpoints following [getting_started.md](../../
...
+
+## IoU loss
+
+Adding IoU loss (simply =1-IoU) boosts VoteNet's performance.
+
+|  Backbone   | Lr schd | Mem (GB) | Inf time (fps) | AP@0.25 |AP@0.5| Download |
+| :---------: | :-----: | :------: | :------------: | :----: |:----: | :------: |
+|    [PointNet++](./votenet_iouloss_8x8_scannet-3d-18class.py)     |  3x    |4.1||63.81|44.21|/|
+
+Please refer to [votenet_iouloss_8x8_scannet-3d-18class.py](./votenet_iouloss_8x8_scannet-3d-18class.py) for the added loss term. Currently we only support calculating IoU loss for axis-aligned bounding boxes since the general IoU calculation CUDA op doesn't support a backward method. Therefore IoU loss is correctly implemented for ScanNet dataset only.
```
Comment (ZwwWayne) [inline configs/votenet/README.md:50]: 1. Simply add a config of that loss for usage in this readme.
2. rephrase
```markdown
For now, we only support calculating IoU loss for axis-aligned bounding boxes since the CUDA op of general 3D IoU calculation does not implement the backward method. Therefore, IoU loss can only be used for ScanNet dataset for now.
```

File: mmdet3d/models/losses/iou_loss_axis_aligned.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
+import torch
+from torch import nn as nn
+
+from mmdet.models.builder import LOSSES
+from mmdet.models.losses.utils import weighted_loss
+
+
+@weighted_loss
+def iou_loss_axis_aligned(pred, target, eps=1e-6):
```
Comment (ZwwWayne) [inline mmdet3d/models/losses/iou_loss_axis_aligned.py:9]: iou_loss_axis_aligned -> axis_aligned_iou_loss

File: mmdet3d/models/losses/iou_loss_axis_aligned.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    union = vol_a + vol_b - intersection
+    eps = union.new_tensor([eps])
+    iou_axis_aligned = intersection / torch.max(union, eps)
+
+    iou_loss = 1 - iou_axis_aligned
+    return iou_loss
+
+
+@LOSSES.register_module()
+class IoULossAxisAligned(nn.Module):
```
Comment (ZwwWayne) [inline mmdet3d/models/losses/iou_loss_axis_aligned.py:39]: IoULossAxisAligned -> AxisAlignedIoULoss

File: mmdet3d/models/dense_heads/vote_head.py [hunk 12]
```diff
@@ -526,10 +554,11 @@ def get_targets_single(self,
         size_res_targets /= pos_mean_sizes
 
         mask_targets = gt_labels_3d[assignment]
+        center_all_targets = center_targets[assignment]
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/vote_head.py:557]: Should we also use assigned_center_targets here for consistency?

Comment (THU17cyz) [reply mmdet3d/models/dense_heads/vote_head.py]: yes, my mistake

File: mmdet3d/models/dense_heads/vote_head.py (hunk 8 above)
Comment (THU17cyz) [reply mmdet3d/models/dense_heads/vote_head.py]: Yes, yapf did this. I managed to avoid this.

File: configs/votenet/votenet_iouloss_8x8_scannet-3d-18class.py [hunk 13]
```diff
@@ -0,0 +1,45 @@
+_base_ = [
```
Comment (ZwwWayne) [inline configs/votenet/votenet_iouloss_8x8_scannet-3d-18class.py:1]: We may directly inherit this config from `votenet_8x8_scannet-3d-18class.py` thus the config only contains the modification by iouloss.

File: mmdet3d/models/losses/axis_aligned_iou_loss.py [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+
+@weighted_loss
+def axis_aligned_iou_loss(pred, target, eps=1e-6):
+    """Calculate the IoU loss (1-IoU) of two set of axis aligned bounding
+    boxes. Note that predictions and targets are one-to-one corresponded.
+
+    Args:
+        pred (torch.Tensor): Bbox predictions with shape [..., 3].
+        target (torch.Tensor): Bbox targets (gt) with shape [..., 3].
+
```
Comment (ZwwWayne) [inline mmdet3d/models/losses/axis_aligned_iou_loss.py:16]: eps (float, optional): xxxxx

File: mmdet3d/models/losses/axis_aligned_iou_loss.py [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+    Returns:
+        torch.Tensor: IoU loss between predictions and targets.
+    """
+
+    (min_a, max_a) = torch.split(pred, 3, dim=-1)
```
Comment (ZwwWayne) [inline mmdet3d/models/losses/axis_aligned_iou_loss.py:21]: It seems that line 21-32 can be abstract as `axis_aligned_iou`, we can do that thus this IoU function can be used in more places.

**Comment by THU17cyz** [comment]
Abstracted axis-aligned 3D IoU as a IoU calculator as [here](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/core/bbox/iou_calculators/iou2d_calculator.py)

File: mmdet3d/models/dense_heads/vote_head.py [hunk 16]
```diff
@@ -318,6 +324,26 @@ def loss(self,
             size_class_loss=size_class_loss,
             size_res_loss=size_res_loss)
 
+        if self.iou_loss:
+            mean_sizes = size_res_targets.new_tensor(
```
Comment (ZwwWayne) [inline mmdet3d/models/dense_heads/vote_head.py:328]: Can we use `self.bbox_coder.decode()`  and `bboxes.corner` to ease some burden?

File: mmdet3d/core/bbox/iou_calculators/iou3d_calculator.py [hunk 17]
```diff
@@ -163,3 +165,153 @@ def bbox_overlaps_3d(bboxes1, bboxes2, mode='iou', coordinate='camera'):
...
+def axis_aligned_bbox_overlaps_3d(bboxes1,
+                                  bboxes2,
+                                  mode='iou',
+                                  is_aligned=False,
+                                  eps=1e-6):
+    """Calculate overlap between two set of axis aligned 3D bboxes.
+    If ``is_aligned `` is ``False``, then calculate the overlaps between each
+    bbox of bboxes1 and bboxes2, otherwise the overlaps between each aligned
+    pair of bboxes1 and bboxes2.
+    Args:
```
Comment (ZwwWayne) [inline mmdet3d/core/bbox/iou_calculators/iou3d_calculator.py:210]: Empty line

File: mmdet3d/core/bbox/iou_calculators/iou3d_calculator.py [hunk 18]
```diff
@@ -163,3 +165,153 @@ def bbox_overlaps_3d(bboxes1, bboxes2, mode='iou', coordinate='camera'):
...
+            If ``is_aligned `` is ``True``, then m and n must be equal.
+        mode (str): "iou" (intersection over union) or "giou" (generalized
+            intersection over union).
+        is_aligned (bool, optional): If True, then m and n must be equal.
+            Default False.
+        eps (float, optional): A value added to the denominator for numerical
+            stability. Default 1e-6.
+    Returns:
+        Tensor: shape (m, n) if ``is_aligned `` is False else shape (m,)
+    Example:
```
Comment (ZwwWayne) [inline mmdet3d/core/bbox/iou_calculators/iou3d_calculator.py:225]: Empty line above `Example`, `Returns`

**Review by ZwwWayne — APPROVED** [review]
